File: src/main/kotlin/com/xs0/gqlktx/types/gql/GType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import com.xs0.gqlktx.schema.intro.GqlIntroType;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

public abstract class GType {
    private final GqlIntroType introspector;

    protected GType() {
        this.introspector = new GqlIntroType(this);
    }

    public GqlIntroType introspector() {
        return introspector;
    }

    public abstract TypeKind getKind();
    public abstract String getGqlTypeString();

    public abstract boolean validAsArgumentType();
    public abstract boolean validAsQueryFieldType();

    public abstract GBaseType getBaseType();

    private GNotNullType myNotNullType;
    public GNotNullType notNull() {
        if (myNotNullType == null)
            myNotNullType = new GNotNullType(this);

        return myNotNullType;
    }

    private GListType myListType;
    public GListType listOf() {
        if (myListType == null)
            myListType = new GListType(this);

        return myListType;
    }

    public abstract void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException;
    public abstract void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException;
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.dom.Value
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.schema.intro.GqlIntroType

abstract class GType protected constructor() {
    val introspector = GqlIntroType(this)

    abstract val kind: TypeKind
    abstract val gqlTypeString: String

    abstract val validAsArgumentType: Boolean
    abstract val validAsQueryFieldType: Boolean

    abstract val baseType: GBaseType

    private var myNotNullType: GNotNullType? = null
    fun notNull(): GNotNullType {
        var cached = myNotNullType
        if (cached == null) {
            cached = GNotNullType(this)
            myNotNullType = cached
        }
        return cached
    }

    private var myListType: GListType? = null
    fun listOf(): GListType {
        var cached = myListType
        if (cached == null) {
            cached = GListType(this)
            myListType = cached
        }
        return cached
    }

    abstract fun coerceValue(raw: Value): Value
}
